src/braket/reservations/reservations.py

```diff
@@ -0,0 +1,90 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from __future__ import annotations
+
+import os
+from contextlib import AbstractContextManager
+
+from braket.aws import AwsDevice
+from braket.devices import Device
+
+
+class DirectReservation(AbstractContextManager):
+    """
+    Context manager that modifies AwsQuantumTasks created within the context to use a reservation
+    ARN for all tasks targeting the specified device.
+
+    Reservations are AWS account and device specific. Only the AWS account that created the
+    reservation can use your reservation ARN. Additionally, the reservation ARN is only valid on the
+    reserved device at the chosen start and end times.
+
+    Args:
+        device (Device | str): The Braket device for which you have a reservation ARN, or
+            optionally the device ARN.
+        reservation_arn (str | None): The Braket Direct reservation ARN to be applied to all
+            quantum tasks run within the context.
+
+    Examples:
+        As a context manager
+        >>> with DirectReservation(device_arn, reservation_arn="<my_reservation_arn>"):
+        ...     task1 = device.run(circuit, shots)
+        ...     task2 = device.run(circuit, shots)
+
+        or start the reservation
+        >>> DirectReservation(device_arn, reservation_arn="<my_reservation_arn>").start()
+        ... task1 = device.run(circuit, shots)
+        ... task2 = device.run(circuit, shots)
+
+    References:
+
+    [1] https://docs.aws.amazon.com/braket/latest/developerguide/braket-reservations.html
+    """
+
+    def __init__(self, device: Device | str, reservation_arn: str | None):
```

Comment:
I'm not following what checks we are doing in this constructor. The error message in L62 suggests that we wish to be stringent (only AwsDevice, or their ARN, and local simulator Device (?)). Is this true? If so, what is the purpose of allowing arbitrary string? 

Also, why do we fill in `""` for when device is a `Device`, what is the use case for passing a local sim device to a reservation? It's likely a mistake by the user, and we should help them catch it, no?